### PR TITLE
GPT Review development-guidelines

### PR DESCRIPTION
**Pull request by 0xPhaze**

### CODE REVIEW

File: development-guidelines/guidelines.md [hunk 1]
```diff
@@ -3,113 +3,113 @@
...
+## Design Guidelines
 
-The design of the contract should be discussed ahead of time, prior writing any line of code.
+Discuss the design of the contract ahead of time, before writing any code.
 
-### Documentation and specifications
+### Documentation and Specifications
 
-Documentation can be written at different levels, and should be updated while implementing the contracts:
+Update documentation at different levels as you implement the contracts:
```
Comment (tuturu-tech) [inline development-guidelines/guidelines.md:27]:
```suggestion
Write documentation at different levels and update it as you implement the contracts:
```

File: development-guidelines/guidelines.md [hunk 2]
```diff
@@ -3,113 +3,113 @@
...
 
-- **A plain English description of the system**, describing what the contracts do and any assumptions on the codebase.
-- **Schema and architectural diagrams**, including the contract interactions and the state machine of the system. [Slither printers](https://github.com/crytic/slither/wiki/Printer-documentation) can help to generate these schemas.
-- **Thorough code documentation**, the [Natspec format](https://solidity.readthedocs.io/en/develop/natspec-format.html) can be used for Solidity.
+- **A plain English description of the system**, describing the contracts' purpose and any assumptions about the codebase.
+- **Schema and architectural diagrams**, including contract interactions and the system's state machine. Use [Slither printers](https://github.com/crytic/slither/wiki/Printer-documentation) to generate these schemas.
```
Comment (tuturu-tech) [inline development-guidelines/guidelines.md:30]:
```suggestion
- **Schema and architectural diagrams**, including contract interactions and the system's state machine. Use [Slither printers](https://github.com/crytic/slither/wiki/Printer-documentation) to help generate these schemas.
```

File: development-guidelines/guidelines.md [hunk 3]
```diff
@@ -3,113 +3,113 @@
...
+  - The keys' storage location and access method
+  - Deployment verification: develop and test a post-deployment script.
 
 #### Delegatecall Proxy Pattern
 
-The delegatecall opcode is a very sharp tool that must be used carefully. Many high-profile exploits utilize little-known edge cases and counter-intuitive aspects of the delegatecall proxy pattern. This section aims to outline the most important risks to keep in mind while developing such smart contract systems. Trail of Bits developed the [slither-check-upgradability](https://github.com/crytic/slither/wiki/Upgradeability-Checks) tool to aid in the development of secure delegatecall proxies, it performs safety checks relevant to both upgradable and immutable delegatecall proxies.
+The delegatecall opcode is a sharp tool that must be used carefully. Many high-profile exploits involve little-known edge cases and counter-intuitive aspects of the delegatecall proxy pattern. To develop secure delegatecall proxies, utilize the [slither-check-upgradability](https://github.com/crytic/slither/wiki/Upgradeability-Checks) tool, which performs safety checks for both upgradable and immutable delegatecall proxies.
```
Comment (tuturu-tech) [inline development-guidelines/guidelines.md:50]:
```suggestion
The delegatecall opcode is a sharp tool that must be used carefully. Many high-profile exploits involve little-known edge cases and counter-intuitive aspects of the delegatecall proxy pattern. To aid the development of secure delegatecall proxies, utilize the [slither-check-upgradability](https://github.com/crytic/slither/wiki/Upgradeability-Checks) tool, which performs safety checks for both upgradable and immutable delegatecall proxies.
```

File: development-guidelines/guidelines.md [hunk 4]
```diff
@@ -3,113 +3,113 @@
...
+- **Storage layout**: Proxy and implementation storage layouts must be the same. Instead of defining the same state variables for each contract, both should inherit all state variables from a shared base contract.
+- **Inheritance**: Be aware that the order of inheritance affects the final storage layout. For example, `contract A is B, C` and `contract A is C, B` will not have the same storage layout if both B and C define state variables.
+- **Initialization**: Immediately initialize the implementation. Implement an uninitialized implementation contract to prevent well-known disasters (and near disasters). A factory pattern can help ensure correct deployment and initialization and reduce front-running risks.
```
Comment (tuturu-tech) [inline development-guidelines/guidelines.md:54]:
```suggestion
- **Initialization**: Immediately initialize the implementation. Well-known disasters (and near disasters) have featured an uninitialized implementation contract. A factory pattern can help ensure correct deployment and initialization and reduce front-running risks.
```

File: development-guidelines/guidelines.md [hunk 5]
```diff
@@ -3,113 +3,113 @@
...
+- **Function shadowing**: Ensure the proxy’s function is called if the same method is defined on the proxy and implementation. Be mindful of `setOwner` and other administration functions commonly found on proxies.
```
Comment (tuturu-tech) [inline development-guidelines/guidelines.md:55]:
```suggestion
- **Function shadowing**: If the same method is defined on the proxy and the implementation, then the proxy’s function will not be called. Be mindful of `setOwner` and other administration functions commonly found on proxies.
```

File: development-guidelines/guidelines.md [hunk 6]
```diff
@@ -3,113 +3,113 @@
...
+- **Direct implementation usage**: Configure implementation state variables with values that prevent direct use, such as setting a flag during construction that disables the implementation and causes all methods to revert. This is particularly important if the implementation also carries delegatecall operations, as this may lead to accidental self-destruction.
```
Comment (tuturu-tech) [inline development-guidelines/guidelines.md:56]:
```suggestion
- **Direct implementation usage**: Configure implementation state variables with values that prevent direct use, such as setting a flag during construction that disables the implementation and causes all methods to revert. This is particularly important if the implementation also performs delegatecall operations, as this may lead to unintended self-destruction of the implementation.
```

File: development-guidelines/guidelines.md [hunk 7]
```diff
@@ -3,113 +3,113 @@
...
+- **Immutable and constant variables**: These variables are embedded in the bytecode and can get out of sync between the proxy and implementation. The implementation may still use an incorrect immutable variable, even if the proxy’s bytecode correctly sets the same variables.
```
Comment (tuturu-tech) [inline development-guidelines/guidelines.md:57]:
```suggestion
- **Immutable and constant variables**: These variables are embedded in the bytecode and can get out of sync between the proxy and implementation. If the implementation has an incorrect immutable variable, this value may still be used even if the same variable is correctly set in the proxy’s bytecode.
```

File: development-guidelines/guidelines.md [hunk 8]
```diff
@@ -3,113 +3,113 @@
...
-- [Contract Upgrade Risks and Recommendations](https://youtu.be/mebA5Qz9zeQ?t=353): A talk by Trail of Bits describing best-practices for developing upgradable delegatecall proxies. The section starting at 5:49 describes some general risks that also apply to non-upgradable proxies.
+- [Contract Upgradability Anti-Patterns](https://blog.trailofbits.com/2018/09/05/contract-upgrade-anti-patterns/): Describes the differences between downstream data contracts and delegatecall proxies with upstream data contracts and how these patterns affect upgradability.
+- [How the Diamond Standard Falls Short](https://blog.trailofbits.com/2020/10/30/good-idea-bad-design-how-the-diamond-standard-falls-short/): Explores delegatecall risks that apply to all contracts, not just those following the diamond standard.
+- [Breaking Aave Upgradeability](https://blog.trailofbits.com/2020/12/16/breaking-aave-upgradeability/): Discusses a subtle problem discovered in Aave `AToken` contracts, resulting from the interplay between delegatecall proxies, contract existence checks, and unsafe initialization.
```
Comment (tuturu-tech) [inline development-guidelines/guidelines.md:64]:
```suggestion
- [Breaking Aave Upgradeability](https://blog.trailofbits.com/2020/12/16/breaking-aave-upgradeability/): Discusses a subtle problem we discovered in Aave `AToken` contracts, resulting from the interplay between delegatecall proxies, contract existence checks, and unsafe initialization.
```

File: development-guidelines/guidelines.md [hunk 9]
```diff
@@ -3,113 +3,113 @@
...
+- [Contract Upgrade Risks and Recommendations](https://youtu.be/mebA5Qz9zeQ?t=353): A Trail of Bits talk on best practices for developing upgradable delegatecall proxies. The section starting at 5:49 describes general risks for non-upgradable proxies.
 
-## Implementation guidelines
+## Implementation Guidelines
 
-**Strive for simplicity.** Always use the simplest solution that fits your purpose. Any member of your team should be able to understand your solution.
+**Aim for simplicity.** Use the simplest solution that meets your needs. Any team member should understand your solution.
```
Comment (tuturu-tech) [inline development-guidelines/guidelines.md:69]:
```suggestion
**Aim for simplicity.** Use the simplest solution that meets your needs. Any member of your team should understand your solution.
```

File: development-guidelines/guidelines.md [hunk 10]
```diff
@@ -3,113 +3,113 @@
...
-- **Split the logic of your system**, either through multiple contracts or by grouping similar functions together (for example, authentication, arithmetic, ...).
-- **Write small functions, with a clear purpose.** This will facilitate easier review and allow the testing of individual components.
+- **Divide the system's logic**, either through multiple contracts or by grouping similar functions together (e.g. authentication, arithmetic).
+- **Write small functions with clear purposes.**
 
 ### Inheritance
 
-- **Keep the inheritance manageable.** Inheritance should be used to divide the logic, however, your project should aim to minimize the depth and width of the inheritance tree.
-- **Use Slither’s [inheritance printer](https://github.com/crytic/slither/wiki/Printer-documentation#inheritance-graph) to check the contracts’ hierarchy.** The inheritance printer will help you review the size of the hierarchy.
+- **Keep inheritance manageable.** Though inheritance can help divide logic, minimize the depth and width of the inheritance tree.
```
Comment (tuturu-tech) [inline development-guidelines/guidelines.md:80]:
```suggestion
- **Keep inheritance manageable.** Though inheritance can help divide logic you should aim to minimize the depth and width of the inheritance tree.
```

File: development-guidelines/guidelines.md [hunk 11]
```diff
@@ -3,113 +3,113 @@
...
-- **Secure the wallets of privileged users.** Follow our [best practices](https://blog.trailofbits.com/2018/11/27/10-rules-for-the-secure-use-of-cryptocurrency-hardware-wallets/) if you store keys in hardware wallets.
-- **Have a response to incident plan.** Consider that your smart contracts can be compromised. Even if your contracts are free of bugs, an attacker may take control of the contract owner's keys.
+- **Monitor contracts.** Observe logs and be prepared to respond in the event of contract or wallet compromise.
+- **Add contact information to [blockchain-security-contacts](https://github.com/crytic/blockchain-security-contacts).** This list helps third parties notify you of discovered security flaws.
+- **Secure privileged users' wallets.** Follow our [best practices](https://blog.trailofbits.com/2018/11/27/10-rules-for-the-secure-use-of-cryptocurrency-hardware-wallets/) for hardware wallet key storage.
+- **Develop a response-to-incident plan.** Assume your smart contracts can be compromised. Possible threats include contract bugs or attackers gaining control of the contract owner's keys.
```
Comment (tuturu-tech) [inline development-guidelines/guidelines.md:115]:
```suggestion
- **Develop a incident response plan.** Assume your smart contracts can be compromised. Possible threats include contract bugs or attackers gaining control of the contract owner's keys.
```

File: development-guidelines/incident_response.md [hunk 12]
```diff
@@ -1,30 +1,30 @@
 # Incident Response Recommendations
 
-Here, we provide recommendations around the formulation of an incident response plan.
+In this article, we provide recommendations for formulating a robust incident response plan.
```
Comment (tuturu-tech) [inline development-guidelines/incident_response.md:3]:
```suggestion
In this section, we provide recommendations for formulating a robust incident response plan.
```

File: development-guidelines/incident_response.md [hunk 13]
```diff
@@ -1,30 +1,30 @@
...
-  - Document the intervals at which the team should perform dry runs of the various scenarios. For scenarios that are more likely to happen, perform dry runs more regularly. Create a template to be filled in after a dry run to describe the improvements that need to be made to the incident response.
+- [ ] **Identify specific individuals or roles responsible for carrying out the mitigations (deploying smart contracts, pausing contracts, upgrading the front end, etc.).**
+  - Defining these roles will enhance the incident response plan and facilitate the execution of mitigation actions when necessary.
+- [ ] **Document internal processes in cases where deployed remediation fails or introduces new bugs.**
+  - Consider developing a fallback plan that outlines an action strategy for failed remediation attempts.
+- [ ] **Provide a clear description of the intended contract deployment process.**
+- [ ] **Contemplate whether and under what circumstances your company will compensate affected users in the event of certain issues.**
```
Comment (tuturu-tech) [inline development-guidelines/incident_response.md:10]:
```suggestion
- [ ] **Consider whether and under what circumstances your company will compensate affected users in the event of certain issues.**
```

File: development-guidelines/incident_response.md [hunk 14]
```diff
@@ -1,30 +1,30 @@
...
+  - Some situations to consider include individual or aggregate losses, losses resulting from user error, contract flaws, and third-party contract defects.
```
Comment (tuturu-tech) [inline development-guidelines/incident_response.md:11]:
```suggestion
  - Some situations to consider include individual or aggregate losses, losses resulting from user error, contract flaws, and third-party contract flaws.
```

File: development-guidelines/incident_response.md [hunk 15]
```diff
@@ -1,30 +1,30 @@
...
+- [ ] **Outline a plan for staying informed about new issues, so as to ensure future development and the security of the deployment toolchain and the external on-chain and off-chain services your system depends on.**
```
Comment (tuturu-tech) [inline development-guidelines/incident_response.md:12]:
```suggestion
- [ ] **Outline a plan for staying informed about new issues, so as to inform future development and enhance the security of the deployment toolchain and the external on-chain and off-chain services your system depends on.**
```

File: development-guidelines/incident_response.md [hunk 16]
```diff
@@ -1,30 +1,30 @@
...
+- [ ] **Conduct periodic dry runs of specific scenarios in the incident response plan to identify gaps and improvement opportunities, and build muscle memory.**
+  - Establish intervals for performing dry runs for each scenario. Conduct more frequent dry runs for scenarios with higher likelihoods of occurrence. Create a template to document improvements required after each dry run for the incident response plan.
 
 ## Incident Response Plan Resources
 
 - [How to Hack the Yield Protocol](https://docs.yieldprotocol.com/#/operations/how_to_hack)
 - [Emergency Steps – Yearn](https://github.com/yearn/yearn-devdocs/blob/master/docs/developers/v2/EMERGENCY.md)
 
-## Well-handled IR Incidents
+## Examples of Well-Handled Incident Response Incidents
```
Comment (tuturu-tech) [inline development-guidelines/incident_response.md:28]:
```suggestion
## Examples of Well-Handled Incidents
```

File: development-guidelines/token_integration.md [hunk 17]
```diff
@@ -1,15 +1,15 @@
-# Token integration checklist
+# Token Integration Checklist
 
-The following checklist provides recommendations for interactions with arbitrary tokens. Every unchecked item should be justified, and its associated risks, understood.
+This checklist offers recommendations for engaging with arbitrary tokens. Ensure that every unchecked item is justified and that its risks are understood.
```
Comment (tuturu-tech) [inline development-guidelines/token_integration.md:3]:
```suggestion
This checklist offers recommendations for interacting with arbitrary tokens. Ensure that every unchecked item is justified and that its risks are understood.
```

File: development-guidelines/token_integration.md [hunk 18]
```diff
@@ -18,79 +18,78 @@ To follow this checklist, use the below output from Slither for the token:
 - slither-prop . --contract ContractName # requires configuration, and use of Echidna and Manticore
 ```
 
-## General considerations
+## General Considerations
 
-- [ ] **The contract has a security review.** Avoid interacting with contracts that lack a security review. Check the length of the assessment (i.e., the level of effort), the reputation of the security firm, and the number and severity of the findings.
-- [ ] **You have contacted the developers.** You may need to alert their team to an incident. Look for appropriate contacts on [blockchain-security-contacts](https://github.com/crytic/blockchain-security-contacts).
-- [ ] **They have a security mailing list for critical announcements.** Their team should advise users (like you!) when critical issues are found or when upgrades occur.
+- [ ] **The contract has a security review.** Avoid engaging with contracts that lack a security review. Assess the review's duration (i.e., the level of effort), the reputation of the security firm, and the number and severity of findings.
```
Comment (tuturu-tech) [inline development-guidelines/token_integration.md:23]:
```suggestion
- [ ] **The contract has a security review.** Avoid interacting with contracts that lack a security review. Assess the review's duration (i.e., the level of effort), the reputation of the security firm, and the number and severity of findings.
```

File: development-guidelines/token_integration.md [hunk 19]
```diff
@@ -18,79 +18,78 @@ To follow this checklist, use the below output from Slither for the token:
...
-- [ ] **The token is not pausable.** Malicious or compromised owners can trap contracts relying on pausable tokens. Identify pausable code by hand.
-- [ ] **The owner cannot blacklist the contract.** Malicious or compromised owners can trap contracts relying on tokens with a blacklist. Identify blacklisting features by hand.
-- [ ] **The team behind the token is known and can be held responsible for abuse.** Contracts with anonymous development teams or teams that reside in legal shelters require a higher standard of review.
+- [ ] **The token is not upgradeable.** Upgradeable contracts may change their rules over time. Use Slither’s [`human-summary` printer](https://github.com/crytic/slither/wiki/Printer-documentation#contract-summary) to determine if the contract is upgradeable.
+- [ ] **The owner has limited minting capabilities.** Malicious or compromised owners can abuse minting capabilities. Use Slither’s [`human-summary` printer](https://github.com/crytic/slither/wiki/Printer-documentation#contract-summary) to review minting capabilities and consider manual code review.
```
Comment (tuturu-tech) [inline development-guidelines/token_integration.md:37]:
```suggestion
- [ ] **The owner has limited minting capabilities.** Malicious or compromised owners can abuse minting capabilities. Use Slither’s [`human-summary` printer](https://github.com/crytic/slither/wiki/Printer-documentation#contract-summary) to review minting capabilities and consider manually reviewing the code.
```

File: development-guidelines/token_integration.md [hunk 20]
```diff
@@ -18,79 +18,78 @@ To follow this checklist, use the below output from Slither for the token:
...
 - [ ] **The `name`, `decimals`, and `symbol` functions are present if used.** These functions are optional in the ERC721 standard and may not be present.
-- [ ] **If it is used, `decimals` returns a `uint8(0)`.** Other values are invalid.
+- [ ] **If used, `decimals` returns a `uint8(0)`.** Other values are invalid.
 - [ ] **The `name` and `symbol` functions can return an empty string.** This behavior is allowed by the standard.
-- [ ] **The `ownerOf` function reverts if the `tokenId` is invalid or is set to a token that has already been burned.** The function cannot return 0x0. This behavior is required by the standard, but it is not always properly implemented.
+- [ ] **The `ownerOf` function reverts if the `tokenId` is invalid or refers to a token that has already been burned.** The function cannot return 0x0. This behavior is required by the standard but may not always be implemented correctly.
 - [ ] **A transfer of an NFT clears its approvals.** This is required by the standard.
-- [ ] **The token ID of an NFT cannot be changed during its lifetime.** This is required by the standard.
```
Comment (tuturu-tech) [inline development-guidelines/token_integration.md:88]: Revert deletion if unintentional

File: development-guidelines/workflow.md [hunk 21]
```diff
@@ -1,43 +1,43 @@
-# Secure development workflow
+# Secure Development Workflow
 
-Here's a high-level process we recommend following while you write your smart contracts.
+Follow this high-level process while writing your smart contracts for enhanced security:
```
Comment (tuturu-tech) [inline development-guidelines/workflow.md:3]:
```suggestion
Follow this high-level process while developing your smart contracts for enhanced security:
```

File: development-guidelines/workflow.md [hunk 22]
```diff
@@ -1,43 +1,43 @@
...
 
-Check for known security issues:
+1. **Check for known security issues:**
 
-- [ ] Review your contracts with [Slither](https://github.com/crytic/slither). It has more than 70 built-in detectors for common vulnerabilities. Run it on every check-in with new code and ensure it gets a clean report (or use triage mode to silence certain issues).
+- Review your contracts using [Slither](https://github.com/crytic/slither), which has over 70 built-in detectors for common vulnerabilities. Run it on every check-in with new code and ensure it gets a clean report (or use triage mode to silence certain issues).
```
Comment (tuturu-tech) [inline development-guidelines/workflow.md:7]:
```suggestion
- [ ] Review your contracts using [Slither](https://github.com/crytic/slither), which has over 70 built-in detectors for common vulnerabilities. Run it on every check-in with new code and ensure it gets a clean report (or use triage mode to silence certain issues).
```

File: development-guidelines/workflow.md [hunk 23]
```diff
@@ -1,43 +1,43 @@
...
-- [ ] Do your contracts purport to conform to ERCs? Check them with [`slither-check-erc`](https://github.com/crytic/slither/wiki/ERC-Conformance). This tool instantly identifies deviations from six common specs.
-- [ ] Do you have unit tests in Truffle? Enrich them with [`slither-prop`](https://github.com/crytic/slither/wiki/Property-generation). It automatically generates a robust suite of security properties for features of ERC20 based on your specific code.
-- [ ] Do you integrate with 3rd party tokens? Review our [token integration checklist](./token_integration.md) before relying on external contracts.
+- If your contracts are upgradeable, review your upgradeability code for flaws using [`slither-check-upgradeability`](https://github.com/crytic/slither/wiki/Upgradeability-Checks) or [Crytic](https://blog.trailofbits.com/2020/06/12/upgradeable-contracts-made-safer-with-crytic/). We have documented 17 ways upgrades can go sideways.
```
Comment (tuturu-tech) [inline development-guidelines/workflow.md:11]:
```suggestion
- [ ] If your contracts are upgradeable, review your upgradeability code for flaws using [`slither-check-upgradeability`](https://github.com/crytic/slither/wiki/Upgradeability-Checks) or [Crytic](https://blog.trailofbits.com/2020/06/12/upgradeable-contracts-made-safer-with-crytic/). We have documented 17 ways upgrades can go sideways.
```

File: development-guidelines/workflow.md [hunk 24]
```diff
@@ -1,43 +1,43 @@
...
+- If your contracts claim to conform to ERCs, check them with [`slither-check-erc`](https://github.com/crytic/slither/wiki/ERC-Conformance). This tool instantly identifies deviations from six common specs.
```
Comment (tuturu-tech) [inline development-guidelines/workflow.md:12]:
```suggestion
- [ ] If your contracts claim to conform to ERCs, check them with [`slither-check-erc`](https://github.com/crytic/slither/wiki/ERC-Conformance). This tool instantly identifies deviations from six common specs.
```

File: development-guidelines/workflow.md [hunk 25]
```diff
@@ -1,43 +1,43 @@
...
+- If you have unit tests in Truffle, enrich them with [`slither-prop`](https://github.com/crytic/slither/wiki/Property-generation). It automatically generates a robust suite of security properties for features of ERC20 based on your specific code.
```
Comment (tuturu-tech) [inline development-guidelines/workflow.md:13]:
```suggestion
- [ ] If you have unit tests in Truffle, enrich them with [`slither-prop`](https://github.com/crytic/slither/wiki/Property-generation). It automatically generates a robust suite of security properties for features of ERC20 based on your specific code.
```

File: development-guidelines/workflow.md [hunk 26]
```diff
@@ -1,43 +1,43 @@
...
+- If you integrate with third-party tokens, review our [token integration checklist](./token_integration.md) before relying on external contracts.
```
Comment (tuturu-tech) [inline development-guidelines/workflow.md:14]:
```suggestion
- [ ] If you integrate with third-party tokens, review our [token integration checklist](./token_integration.md) before relying on external contracts.
```

File: development-guidelines/workflow.md [hunk 27]
```diff
@@ -1,43 +1,43 @@
...
 
-Visually inspect critical security features of your code:
+3. **Visually inspect critical security features of your code:**
 
-- [ ] Review Slither's [inheritance-graph](https://github.com/trailofbits/slither/wiki/Printer-documentation#inheritance-graph) printer. Avoid inadvertent shadowing and C3 linearization issues.
-- [ ] Review Slither's [function-summary](https://github.com/trailofbits/slither/wiki/Printer-documentation#function-summary) printer. It reports function visibility and access controls.
-- [ ] Review Slither's [vars-and-auth](https://github.com/trailofbits/slither/wiki/Printer-documentation#variables-written-and-authorization) printer. It reports access controls on state variables.
+- Review Slither's [inheritance-graph](https://github.com/trailofbits/slither/wiki/Printer-documentation#inheritance-graph) printer to avoid inadvertent shadowing and C3 linearization issues.
```
Comment (tuturu-tech) [inline development-guidelines/workflow.md:18]:
```suggestion
- [ ] Review Slither's [inheritance-graph](https://github.com/trailofbits/slither/wiki/Printer-documentation#inheritance-graph) printer to avoid inadvertent shadowing and C3 linearization issues.
```

File: development-guidelines/workflow.md [hunk 28]
```diff
@@ -1,43 +1,43 @@
...
+- Review Slither's [function-summary](https://github.com/trailofbits/slither/wiki/Printer-documentation#function-summary) printer, which reports function visibility and access controls.
```
Comment (tuturu-tech) [inline development-guidelines/workflow.md:19]:
```suggestion
- [ ] Review Slither's [function-summary](https://github.com/trailofbits/slither/wiki/Printer-documentation#function-summary) printer, which reports function visibility and access controls.
```

File: development-guidelines/workflow.md [hunk 29]
```diff
@@ -1,43 +1,43 @@
...
+- Review Slither's [vars-and-auth](https://github.com/trailofbits/slither/wiki/Printer-documentation#variables-written-and-authorization) printer, which reports access controls on state variables.
```
Comment (tuturu-tech) [inline development-guidelines/workflow.md:20]:
```suggestion
- [ ] Review Slither's [vars-and-auth](https://github.com/trailofbits/slither/wiki/Printer-documentation#variables-written-and-authorization) printer, which reports access controls on state variables.
```

File: development-guidelines/workflow.md [hunk 30]
```diff
@@ -1,43 +1,43 @@
...
 
-Document critical security properties and use automated test generators to evaluate them:
+4. **Document critical security properties and use automated test generators to evaluate them:**
 
-- [ ] Learn to [document security properties for your code](../program-analysis/). It's tough as first, but it's the single most important activity for achieving a good outcome. It's also a prerequisite for using any of the advanced techniques in this tutorial.
-- [ ] Define security properties in Solidity, for use with [Echidna](https://github.com/crytic/echidna) and [Manticore](https://manticore.readthedocs.io/en/latest/verifier.html). Focus on your state machine, access controls, arithmetic operations, external interactions, and standards conformance.
-- [ ] Define security properties with [Slither's Python API](../program-analysis/slither). Focus on inheritance, variable dependencies, access controls, and other structural issues.
+- Learn to [document security properties for your code](../program-analysis/). Although challenging at first, it is the single most important activity for achieving a good outcome. It is also a prerequisite for using any advanced techniques in this tutorial.
```
Comment (tuturu-tech) [inline development-guidelines/workflow.md:24]:
```suggestion
- [ ] Learn to [document security properties for your code](../program-analysis/). Although challenging at first, it is the single most important activity for achieving a good outcome. It is also a prerequisite for using any advanced techniques in this tutorial.
```

File: development-guidelines/workflow.md [hunk 31]
```diff
@@ -1,43 +1,43 @@
...
+- Define security properties in Solidity for use with [Echidna](https://github.com/crytic/echidna) and [Manticore](https://manticore.readthedocs.io/en/latest/verifier.html). Focus on your state machine, access controls, arithmetic operations, external interactions, and standards conformance.
```
Comment (tuturu-tech) [inline development-guidelines/workflow.md:25]:
```suggestion
- [ ] Define security properties in Solidity for use with [Echidna](https://github.com/crytic/echidna) and [Manticore](https://manticore.readthedocs.io/en/latest/verifier.html). Focus on your state machine, access controls, arithmetic operations, external interactions, and standards conformance.
```

File: development-guidelines/workflow.md [hunk 32]
```diff
@@ -1,43 +1,43 @@
...
+- Define security properties with [Slither's Python API](../program-analysis/slither). Concentrate on inheritance, variable dependencies, access controls, and other structural issues.
```
Comment (tuturu-tech) [inline development-guidelines/workflow.md:26]:
```suggestion
- [ ] Define security properties with [Slither's Python API](../program-analysis/slither). Concentrate on inheritance, variable dependencies, access controls, and other structural issues.
```

**Review by tuturu-tech — APPROVED** [review]
